Sprite save load

kchadha opened the pull request:
### Resolves

Resolves LLK/scratch-gui#2030

### Proposed Changes

This pull request adds a top level API for exporting a sprite (in the .sprite3 format). This PR also makes provisions for uploading a sprite (.sprite2 or .sprite3) such that any variable conflicts with the existing project are resolved. Without these changes, uploading a .sprite2 (or .sprite3, but those didn't exist before this PR) into an existing project would fail to load the blocks in the sprite with an error in the console in the following scenarios:

#### Scenario 1 - Global Var Reference Conflicting with Project Global Var
- Create a project in Scratch 3 with a global variable named 'foo'
- Create a new project in Scratch 2 with a global variable named 'foo', and a block in a Sprite1's workspace referencing the variable 'foo'.
- Export 'Sprite1' from the Scratch 2 project
- Upload the exported .sprite2 file into the Scratch 3 project created in the first step.
- Observe errors in the console.

#### Scenario 2 - Local Var Reference Conflicting with Project Global Var
- Create a project in Scratch 3 with a global variable named 'foo'
- Create a new project in Scratch 2 with a local variable named 'foo', and [optionally] a block in a Sprite1's workspace referencing the variable 'foo'.
- Export 'Sprite1' from the Scratch 2 project
- Upload the exported .sprite2 file into the Scratch 3 project created in the first step.
- Observe errors in the console.

### Test Coverage

Added unit tests for the variable conflict resolving functionality.

kchadha committed:
Sprite3 Export/Import.

kchadha committed:
Provide global vars to sb2 import when deserializing a single sprite.

kchadha committed:
Fix up variable references after deserializing a .sprite2 or .sprite3 to avoid conflicts with pre-existing variables.

kchadha committed:
Refactor and move variable reference fixup function into target.

kchadha committed:
Add a test for 'getAllVariableAndListReferences' function.

kchadha committed:
fixUpVariableReferences should handle conflicting local variables that don't have blocks referencing them. Add unit tests for the function.

kchadha committed:
Call fixUpVariableReferences in installTargets, before emitting the workspace update.

kchadha committed:
Remove stray newline.

paulkaplan commented on src/engine/target.js:
```diff
@@ -264,6 +332,103 @@ class Target extends EventEmitter {
+     * variable does exist in the target itself (e.g. it's a local variable in the sprite being uploaded),
+     * then the variable is renamed to distinguish itself from the pre-existing variable.
+     * All blocks that reference the local variable will be updated to use the new name.
+     // * @param {Target} target The new target being uploaded, with potential variable conflicts
```
Confused about the commented out jsdoc here, it looks like it shouldn't need any params, but the unit tests pass in params?

kchadha replied:
Oops, that was the remnant of a refactor I did. Fixed.

paulkaplan commented:
@kchadha there seems to be a confusing mismatch in whether the fixup function takes params or not? 

I'm not completely done reviewing, but wanted to give you the chance to fix that before waiting for the full review

paulkaplan commented:
@kchadha I know you asked me about this before hand, but I think I'm going back on what I said earlier. I'm not sure the behavior is Scenario 1 feels right. I was expecting it to hook up the global variable references to the existing global variable, not create a new __local__ variable with a modified name. I feel like the modified name solution works ok for Scenario 2, but doesn't seem right for Scenario 1. Hooking the variable reference up to an existing variable would enhance shareability of sprites, and would unify the behavior you would expect when sharing blocks between sprites (i.e. sharing global blocks refer the global variable, sharing local blocks create a new local variable)

paulkaplan commented:
Looks like my testing was wrong and that is what is implemented...

kchadha committed:
Clean up calls to fixUpVariableReferences

kchadha committed:
Add optional type param to export sprite API.

kchadha commented:
@paulkaplan I fixed the parameters confusion (it was an accidental remnant of a refactor I had done), and I added the optional parameter to specify the resulting format of the zip (defaulting to blob when not provided).

paulkaplan reviewed (approved):
LG!

paulkaplan commented:
Great, LG
